specs/withdrawals.md

```diff
@@ -0,0 +1,176 @@
+
+1. The `L2OutputTimestamp` is an alias for `uint256`, and MUST be a multiple of the `SUBMISSION_INTERVAL` described
+  in the [L2 Output](./proposals.md#constants) document.
+
+1. The `WithdrawalsRoot` is an alias for a `bytes32` value, corresponding to the [MPT][g-mpt]
+  storage root of the Withdrawals predeploy contract at `0x4200000000000000000000000000000000000015` (described below).
+
+1. The `WithdrawalsRootInclusionProof` proof contains the data necessary to prove that the provided `WithdrawalsRoot` is
+  included in the [SSZ merkleization](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#merkleization)
+  of the `L2Output` as defined in the
+  [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+
+1. A `WithdrawalMessageInclusionProof` is an MPT proof encoded in a struct as follows:
+
+   ```js
+   struct WithdrawalMessageInclusionProof {
+       WithdrawalMessage message;
+       bytes32 l2withdrawalsRoot;
+       bytes memory _key;      // storage key of the withdrawal message commitment
+       bytes memory _value;    // Always bytes32(1) (boolean validity status of the message)
+       bytes memory _proof;    // MPT inclusion proof for the key/value
+   }
+   ```
+1. This document also refers to the `L2Output` type as defined in the
+  [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+
+## Withdrawal verification
+
+In order to verify a `WithdrawalMessage`, the following inputs must be provided:
+
+| Type                              | Name               |
+| --------------------------------- | ------------------ |
+| `WithdrawalMessage`               | `message`          |
+| `L2OutputTimestamp`               | `timestamp`        |
+| `WithdrawalsRootInclusionProof`   | `storageRootProof` |
+| `WithdrawalMessageInclusionProof` | `messageProof`     |
+
+These inputs must satisfy the following conditions:
+
+1. The `timestamp` is at least `FINALIZATION_WINDOW` seconds old.
+1. `OutputOracle.l2Outputs(timestamp)` returns a non-zero value `l2Output`.
+1. The `storageRootProof` is valid for the `l2Output` returned in step 2, according to the SSZ encoding described in the
+   [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+1. The `messageProof` is valid for the provided `message` and `messageProof`
+
+## Withdrawal Flow
+
+**On L2:**
+
+1. An L2 account sends a withdrawal message (and possibly also ETH) to the `Withdrawoooor` predeploy contract.
+   This is a very simple contract that stores a mapping from the hash of the `WithdrawalMessage` as defined above to a
+   boolean value. (`mapping (bytes32 => bool) withdrawalMessages`)
+2. If ETH is being withdrawn, it can eventually be burned by deploying a contract which immediately `SELFDESTRUCT`s.
+   This has the benefit of using an existing EVM mechanism for removing the ETH from the world state, without having to
+   add a diff to L2 Geth.
```

Comment:
ACD has had `SELFDESTRUCT` in its crosshair for a while now, there's been talk of at least refusing new contracts that use it, etc. So this might introduce more complexity down the line.

That being said, maybe we could implement like this now, as long as we have a possible upgrade path (which would simply be the execution engine fishing out these transactions and burning the ETH).

Reply:
Ugh using SELFDESTRUCT upsets me. It is probably preferable to adding an EVM diff though... We should keep our eyes out for a better solution but this is a fine default (until we remove SELFDESTRUCT from the EVM)